File: MorySnip/Classes/Screenshot.vb
Public Class Screenshot
    Inherits MarshalByRefObject

    Public Image As Image
    Public Title As String
    Public Comment As String
    Public OriginalPath As String

    Public Shared Widening Operator CType(source As Image) As Screenshot
        Return New Screenshot With {.Image = source}
    End Operator

    Public Shared Narrowing Operator CType(source As Screenshot) As Image
        Return source.Image
    End Operator

    Public Function Clone() As Screenshot
        Dim Cloned As New Screenshot

        If Me.Image IsNot Nothing Then Cloned.Image = Me.Image.Clone()
        If Me.Comment IsNot Nothing Then Cloned.Comment = Me.Comment.Clone()
        If Me.Title IsNot Nothing Then Cloned.Title = Me.Title.Clone()
        If Me.OriginalPath IsNot Nothing Then Cloned.OriginalPath = Me.OriginalPath.Clone()

        Return Cloned
    End Function

    Public Function GetThumbnailImage() As Bitmap
        Return GetThumbnailImage(128, 128, Color.White, New Pen(Brushes.DarkGray, 2))
    End Function

    Public Function GetThumbnailImage(Width As Integer, Height As Integer, Fill As Color, Optional Border As Pen = Nothing) As Bitmap
        'a holder for the result
        Dim Result As Bitmap = New Bitmap(Width, Height)

        If Image Is Nothing Then
            Return Result
        End If

        ' set the resolutions the same to avoid cropping due to resolution differences
        Result.SetResolution(Image.HorizontalResolution, Image.VerticalResolution)

        Dim x, y, w, h As Integer


        If Image.Width < Result.Width AndAlso Image.Height < Result.Height Then
            w = Image.Width
            h = Image.Height
            x = (Result.Width - w) / 2
            y = (Result.Height - h) / 2
        Else
            Dim r1 As Double = Image.Width / Image.Height
            Dim r2 As Double = Result.Width / Result.Height

            If r1 = r2 Then
                w = Result.Width
                h = Result.Height
                x = 0
                y = 0
            ElseIf r1 > r2 Then
                w = Result.Width
                h = Image.Height * (Result.Width / Image.Width)
                x = 0
                y = (Height - h) / 2
            ElseIf r1 < r2 Then
                w = Image.Width * (Result.Height / Image.Height)
                h = Result.Height
                x = (Result.Width - w) / 2
                y = 0
            End If
        End If

        'use a graphics object to draw the resized image into the bitmap
        With Graphics.FromImage(Result)
            .Clear(Fill)
            'set the resize quality modes to high quality
            .CompositingQuality = System.Drawing.Drawing2D.CompositingQuality.HighQuality
            .InterpolationMode = System.Drawing.Drawing2D.InterpolationMode.HighQualityBicubic
            .SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality
            'draw the image into the target bitmap
            .DrawImage(Image, x, y, w, h)
        End With

        'return the resulting bitmap
        Return Result
    End Function
End Class

File: MorySnip/Classes/Screenshot.cs
using System;
using System.Drawing;
using Microsoft.VisualBasic.CompilerServices;

namespace Morysoft.MorySnip
{
    public class Screenshot
    {
        public Image Image { get; set; }
        public string Title { get; set; }
        public string Comment { get; set; }
        public string OriginalPath { get; set; }

        public static implicit operator Screenshot(Image source) => new() { Image = source };

        public static explicit operator Image(Screenshot source) => source.Image;

        public Screenshot Clone() => new()
        {
            Image = (Image)this.Image?.Clone(),
            Comment = this.Comment,
            Title = this.Title,
            OriginalPath = this.OriginalPath
        };

        public Bitmap GetThumbnailImage() => this.GetThumbnailImage(128, 128, Color.White);

        public Bitmap GetThumbnailImage(int width, int height, Color fill)
        {
            // a holder for the result
            var result = new Bitmap(width, height);

            if (this.Image == null)
            {
                return result;
            }

            // set the resolutions the same to avoid cropping due to resolution differences
            result.SetResolution(this.Image.HorizontalResolution, this.Image.VerticalResolution);

            int x = default, y = default, w = default, h = default;

            if (this.Image.Width < result.Width && this.Image.Height < result.Height)
            {
                w = this.Image.Width;
                h = this.Image.Height;
                x = Conversions.ToInteger((result.Width - w) / (double)2);
                y = Conversions.ToInteger((result.Height - h) / (double)2);
            }
            else
            {
                double r1 = this.Image.Width / (double)this.Image.Height;
                double r2 = result.Width / (double)result.Height;

                if (r1 == r2)
                {
                    w = result.Width;
                    h = result.Height;
                    x = 0;
                    y = 0;
                }
                else if (r1 > r2)
                {
                    w = result.Width;
                    h = Conversions.ToInteger(this.Image.Height * (result.Width / (double)this.Image.Width));
                    x = 0;
                    y = Conversions.ToInteger((height - h) / (double)2);
                }
                else if (r1 < r2)
                {
                    w = Conversions.ToInteger(this.Image.Width * (result.Height / (double)this.Image.Height));
                    h = result.Height;
                    x = Conversions.ToInteger((result.Width - w) / (double)2);
                    y = 0;
                }
            }

            // use a graphics object to draw the resized image into the bitmap
            {
                var withBlock = Graphics.FromImage(result);

                withBlock.Clear(fill);
                // set the resize quality modes to high quality
                withBlock.CompositingQuality = System.Drawing.Drawing2D.CompositingQuality.HighQuality;
                withBlock.InterpolationMode = System.Drawing.Drawing2D.InterpolationMode.HighQualityBicubic;
                withBlock.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality;
                // draw the image into the target bitmap
                withBlock.DrawImage(this.Image, x, y, w, h);
            }

            // return the resulting bitmap
            return result;
        }
    }
}
